### ISSUE TITLE
model.generate() in training phrase

### ISSUE THREAD

**Issue by gjc0824** [chunhuizng/mllm-video-captioner#1]
Thank you for your highly practical work. However, while reading the code, I encountered some confusion. When performing SCST training on BLIP2_OPT, you used `self.opt_model.generate()` in lavis/models/blip2_models/blip2_opt.py L248 to generate sentences and obtain logit scores. However, in the implementation of the `transformers` library, `model.generate()` is decorated with `@torch.no_grad()`. Therefore, is your implementation capable of performing backpropagation, or have you made any other modifications?

It's also possible that I may have misunderstood something. I look forward to your clarification.

**Comment by chunhuizng** [comment]
> Thank you for your highly practical work. However, while reading the code, I encountered some confusion. When performing SCST training on BLIP2_OPT, you used `self.opt_model.generate()` in lavis/models/blip2_models/blip2_opt.py L248 to generate sentences and obtain logit scores. However, in the implementation of the `transformers` library, `model.generate()` is decorated with `@torch.no_grad()`. Therefore, is your implementation capable of performing backpropagation, or have you made any other modifications?
> 
> It's also possible that I may have misunderstood something. I look forward to your clarification.

Hi, thanks for asking. Yes, we modified the original code built in the Transformers library. Simply using our repository is not enough. Please temporarily remove the @torch.no_grad() in transformers library's model.generate(). Thank you!

**Comment by chunhuizng** [comment]
> Thank you for your highly practical work. However, while reading the code, I encountered some confusion. When performing SCST training on BLIP2_OPT, you used `self.opt_model.generate()` in lavis/models/blip2_models/blip2_opt.py L248 to generate sentences and obtain logit scores. However, in the implementation of the `transformers` library, `model.generate()` is decorated with `@torch.no_grad()`. Therefore, is your implementation capable of performing backpropagation, or have you made any other modifications?
> 
> It's also possible that I may have misunderstood something. I look forward to your clarification.

Or, I guess you could consider another implementation choice, without the change to the transformer package. For example:

1. Roll out (infer, or generate; anyway, they have similar meaning here) a caption (similar to the GRPO's rollout/sampling period).
2. Then, use the rollout results to have the LLM forward them, and obtain the likelihood of the rollout's caption from the logits.
3. Optimize the corresponding policy gradient.

I think this approach would be more scalable if you intend to use GRPO-style reinforcement learning for this tasks.
